Iterator simplification

swooster opened the pull request:
This has a bunch of minor iterator-related spot refactors, primarily to reduce nesting of iterators.

swooster commented on generator/src/lib.rs:
```diff
@@ -17,6 +17,21 @@ use std::fmt::Display;
 use std::path::Path;
 use syn::Ident;
 
+macro_rules! match_variant {
```
I wasn't really sure what to call this... other names I thought of were `partial_match` or `extract_variant`.

Ralith commented on generator/src/lib.rs:
```diff
@@ -1362,22 +1360,21 @@ pub fn generate_enum<'a>(
+            quote! {
+                Self::#variant_ident => Some(#notation)
+            }
+        });
```
I'm not sure this is an improvement; we're replacing a single `filter_map` with a chained `filter_map` -> `map`. It only saves one line of code, and that can be attributed to different formatting of the `EnumsChild` match--in fact, the same syntax tweak would result in the original form being shorter still.

swooster replied:
The same comment as below applies to this... the idea of `get_variant` hadn't occurred to me when I started, so I was initially focusing on trying to unnest things and break them into small clear repeated steps. I'm thinking of redoing the commits so that I define and use `get_variant` first, then unnest in a later commit (I was thinking that first shrinking the code with `get_variant` might also making it easier to see that the unnesting doesn't change behavior).

swooster replied:
It hit me that maybe I should describe why splitting up the code this way appealed to me: It's easier for me to interpret XML-related iters if all intermediate steps are constrained to forms `.filter_map(get_variant!(Foo::BarVariant))` or `.flat_map(&foo.children)` and only the last step has custom logic. Doing that allows me to read the iterator definition as a path: "traverse into `BarVariant` elements, then traverse into their children, then... etc... until we've reached the elements we want to process, then we apply unique application logic to those elements"

swooster replied:
I'll go ahead leave this change out of the latest PR.

Ralith commented on generator/src/lib.rs:
```diff
@@ -2085,22 +2082,14 @@ pub fn generate_feature<'a>(
+            vkxml::FeatureReference::CommandReference(ref cmd_ref) => Some(cmd_ref),
+            _ => None,
         })
         .filter_map(|cmd_ref| commands.get(&cmd_ref.name))
```
Let's not have two `filter_map`s chained in a row.

swooster replied:
Part of this is that it sort of separated out a repeated pattern of `match x { X::Foo(y) => y, _ => None }`, which ended up making it clearer to me that a `get_variant` macro would be useful. Would it be preferable to avoid this intermediate commit?

Ralith commented on generator/src/lib.rs:
```diff
@@ -17,6 +17,21 @@ use std::fmt::Display;
 use std::path::Path;
 use syn::Ident;
 
+macro_rules! maybe_extract {
```
Would `get_variant` be a clearer name for this?

swooster replied:
I wasn't sure what name was clearest and had tried a bunch of different ones... `get_variant` sounds good to me; I'll switch to that.

swooster commented:
@Ralith: I've pushed a new version... I've renamed `maybe_extract!()` to `get_variant!()` and I've squashed the unnest/flatten/maybe_extract commits together because they tend to go together and the intermediate steps aren't always a clear win. If you preferred those separated, let me know and I'll undo that.

I've undone the change refactoring `generate_result()` per your review, but have left in the change to `generate_feature()` because that allowed collapsing the code down with `get_variant!()`. If you still feel strongly about `generate_feature()` even with `get_variant!()` applied, please let me know.

Also, I changed `.map(|union| union.name.as_str())` to `.map(|union_| union_.name.as_str())` to avoid a binding with a reserved name.

Ralith reviewed (approved):
Good stuff, thanks!
